我在容器中运行frpc,当容器停止后,无法再次启动frpc,除非我重启整个系统

### Bug Description

root@17c36744a7be:/data/docker/frpc# ./frpc -c frpc.toml 
2023/12/20 15:05:10 [I] [root.go:141] start frpc service for config file [frpc.toml]
2023/12/20 15:05:10 [I] [service.go:288] try to connect to server...
2023/12/20 15:05:14 [I] [service.go:279] [48de1c586b168eb3] login to server success, get run id [48de1c586b168eb3]
2023/12/20 15:05:14 [I] [proxy_manager.go:173] [48de1c586b168eb3] proxy added: [portal]
2023/12/20 15:05:14 [I] [service.go:288] [48de1c586b168eb3] try to connect to server...
2023/12/20 15:05:14 [I] [control.go:169] [48de1c586b168eb3] [portal] start proxy success
2023/12/20 15:05:19 [I] [service.go:279] [48de1c586b168eb3] login to server success, get run id [48de1c586b168eb3]
2023/12/20 15:05:19 [I] [proxy_manager.go:173] [48de1c586b168eb3] proxy added: [portal]
2023/12/20 15:05:19 [I] [panic.go:884] frpc service for config file [frpc.toml] stopped
panic: close of closed channel

goroutine 1 [running]:
github.com/fatedier/frp/client.(*Service).loopLoginUntilSuccess.func1()
	github.com/fatedier/frp/client/service.go:331 +0x517
github.com/fatedier/frp/pkg/util/wait.BackoffUntil(0xc000217a10, {0xd2f4a0, 0xc0000fd710}, 0x1, 0xc0000943c0)
	github.com/fatedier/frp/pkg/util/wait/backoff.go:135 +0x118
github.com/fatedier/frp/client.(*Service).loopLoginUntilSuccess(0xc00017c1c0, 0x2540be400, 0x1)
	github.com/fatedier/frp/client/service.go:336 +0x1e9
github.com/fatedier/frp/client.(*Service).Run(0xc00017c1c0, {0xd35450?, 0xc0000a6000?})
	github.com/fatedier/frp/client/service.go:166 +0x10c
github.com/fatedier/frp/cmd/frpc/sub.startService(0xc000158840, {0xc00012dfb0, 0x1, 0x1}, {0x1182358, 0x0, 0x0}, {0x7ffd0071590d, 0x9})
	github.com/fatedier/frp/cmd/frpc/sub/root.go:159 +0x37d
github.com/fatedier/frp/cmd/frpc/sub.runClient({0x7ffd0071590d, 0x9})
	github.com/fatedier/frp/cmd/frpc/sub/root.go:129 +0x19c
github.com/fatedier/frp/cmd/frpc/sub.glob..func2(0x1147100?, {0xb8a06f?, 0x2?, 0x2?})
	github.com/fatedier/frp/cmd/frpc/sub/root.go:69 +0x4c
github.com/spf13/cobra.(*Command).execute(0x1147100, {0xc0000a2160, 0x2, 0x2})
	github.com/spf13/cobra@v1.8.0/command.go:983 +0xaaa
github.com/spf13/cobra.(*Command).ExecuteC(0x1147100)
	github.com/spf13/cobra@v1.8.0/command.go:1115 +0x425
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/fatedier/frp/cmd/frpc/sub.Execute()
	github.com/fatedier/frp/cmd/frpc/sub/root.go:100 +0x25
main.main()


### frpc Version

0.53.0

### frps Version

0.53.0

### System Architecture

linux/amd64

### Configurations

serverAddr = "zchome.us.to"
serverPort = 14443

[[proxies]]
name = "portal"
type = "tcp"
localIP = "nginx"
localPort = 8070
remotePort = 9070


### Logs

root@17c36744a7be:/data/docker/frpc# ./frpc -c frpc.toml 
2023/12/20 15:05:10 [I] [root.go:141] start frpc service for config file [frpc.toml]
2023/12/20 15:05:10 [I] [service.go:288] try to connect to server...
2023/12/20 15:05:14 [I] [service.go:279] [48de1c586b168eb3] login to server success, get run id [48de1c586b168eb3]
2023/12/20 15:05:14 [I] [proxy_manager.go:173] [48de1c586b168eb3] proxy added: [portal]
2023/12/20 15:05:14 [I] [service.go:288] [48de1c586b168eb3] try to connect to server...
2023/12/20 15:05:14 [I] [control.go:169] [48de1c586b168eb3] [portal] start proxy success
2023/12/20 15:05:19 [I] [service.go:279] [48de1c586b168eb3] login to server success, get run id [48de1c586b168eb3]
2023/12/20 15:05:19 [I] [proxy_manager.go:173] [48de1c586b168eb3] proxy added: [portal]
2023/12/20 15:05:19 [I] [panic.go:884] frpc service for config file [frpc.toml] stopped
panic: close of closed channel

goroutine 1 [running]:
github.com/fatedier/frp/client.(*Service).loopLoginUntilSuccess.func1()
	github.com/fatedier/frp/client/service.go:331 +0x517
github.com/fatedier/frp/pkg/util/wait.BackoffUntil(0xc000217a10, {0xd2f4a0, 0xc0000fd710}, 0x1, 0xc0000943c0)
	github.com/fatedier/frp/pkg/util/wait/backoff.go:135 +0x118
github.com/fatedier/frp/client.(*Service).loopLoginUntilSuccess(0xc00017c1c0, 0x2540be400, 0x1)
	github.com/fatedier/frp/client/service.go:336 +0x1e9
github.com/fatedier/frp/client.(*Service).Run(0xc00017c1c0, {0xd35450?, 0xc0000a6000?})
	github.com/fatedier/frp/client/service.go:166 +0x10c
github.com/fatedier/frp/cmd/frpc/sub.startService(0xc000158840, {0xc00012dfb0, 0x1, 0x1}, {0x1182358, 0x0, 0x0}, {0x7ffd0071590d, 0x9})
	github.com/fatedier/frp/cmd/frpc/sub/root.go:159 +0x37d
github.com/fatedier/frp/cmd/frpc/sub.runClient({0x7ffd0071590d, 0x9})
	github.com/fatedier/frp/cmd/frpc/sub/root.go:129 +0x19c
github.com/fatedier/frp/cmd/frpc/sub.glob..func2(0x1147100?, {0xb8a06f?, 0x2?, 0x2?})
	github.com/fatedier/frp/cmd/frpc/sub/root.go:69 +0x4c
github.com/spf13/cobra.(*Command).execute(0x1147100, {0xc0000a2160, 0x2, 0x2})
	github.com/spf13/cobra@v1.8.0/command.go:983 +0xaaa
github.com/spf13/cobra.(*Command).ExecuteC(0x1147100)
	github.com/spf13/cobra@v1.8.0/command.go:1115 +0x425
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/fatedier/frp/cmd/frpc/sub.Execute()
	github.com/fatedier/frp/cmd/frpc/sub/root.go:100 +0x25
main.main()


### Steps to reproduce

1. 
2. 
3. 
...


### Affected area

- [ ] Docs
- [ ] Installation
- [ ] Performance and Scalability
- [ ] Security
- [ ] User Experience
- [ ] Test and Release
- [ ] Developer Infrastructure
- [ ] Client Plugin
- [ ] Server Plugin
- [ ] Extensions
- [ ] Others

--- a/client/control.go
+++ b/client/control.go
@@ -239,15 +239,15 @@ func (ctl *Control) heartbeatWorker() {
 	// Users can still enable heartbeat feature by setting HeartbeatInterval to a positive value.
 	if ctl.sessionCtx.Common.Transport.HeartbeatInterval > 0 {
 		// send heartbeat to server
-		sendHeartBeat := func() error {
+		sendHeartBeat := func() (bool, error) {
 			xl.Debug("send heartbeat to server")
 			pingMsg := &msg.Ping{}
 			if err := ctl.sessionCtx.AuthSetter.SetPing(pingMsg); err != nil {
 				xl.Warn("error during ping authentication: %v, skip sending ping message", err)
-				return err
+				return false, err
 			}
 			_ = ctl.msgDispatcher.Send(pingMsg)
-			return nil
+			return false, nil
 		}
 
 		go wait.BackoffUntil(sendHeartBeat,

--- a/client/service.go
+++ b/client/service.go
@@ -192,16 +192,16 @@ func (svr *Service) keepControllerWorking() {
 	// the control immediately exits. It is necessary to limit the frequency of reconnection in this case.
 	// The interval for the first three retries in 1 minute will be very short, and then it will increase exponentially.
 	// The maximum interval is 20 seconds.
-	wait.BackoffUntil(func() error {
+	wait.BackoffUntil(func() (bool, error) {
 		// loopLoginUntilSuccess is another layer of loop that will continuously attempt to
 		// login to the server until successful.
 		svr.loopLoginUntilSuccess(20*time.Second, false)
 		if svr.ctl != nil {
 			<-svr.ctl.Done()
-			return errors.New("control is closed and try another loop")
+			return false, errors.New("control is closed and try another loop")
 		}
 		// If the control is nil, it means that the login failed and the service is also closed.
-		return nil
+		return false, nil
 	}, wait.NewFastBackoffManager(
 		wait.FastBackoffOptions{
 			Duration:        time.Second,
@@ -282,9 +282,8 @@ func (svr *Service) login() (conn net.Conn, connector Connector, err error) {
 
 func (svr *Service) loopLoginUntilSuccess(maxInterval time.Duration, firstLoginExit bool) {
 	xl := xlog.FromContextSafe(svr.ctx)
-	successCh := make(chan struct{})
 
-	loginFunc := func() error {
+	loginFunc := func() (bool, error) {
 		xl.Info("try to connect to server...")
 		conn, connector, err := svr.login()
 		if err != nil {
@@ -292,7 +291,7 @@ func (svr *Service) loopLoginUntilSuccess(maxInterval time.Duration, firstLoginE
 			if firstLoginExit {
 				svr.cancel(cancelErr{Err: err})
 			}
-			return err
+			return false, err
 		}
 
 		svr.cfgMu.RLock()
@@ -315,7 +314,7 @@ func (svr *Service) loopLoginUntilSuccess(maxInterval time.Duration, firstLoginE
 		if err != nil {
 			conn.Close()
 			xl.Error("NewControl error: %v", err)
-			return err
+			return false, err
 		}
 		ctl.SetInWorkConnCallback(svr.handleWorkConnCb)
 
@@ -328,8 +327,7 @@ func (svr *Service) loopLoginUntilSuccess(maxInterval time.Duration, firstLoginE
 		svr.ctl = ctl
 		svr.ctlMu.Unlock()
 
-		close(successCh)
-		return nil
+		return true, nil
 	}
 
 	// try to reconnect to server until success
@@ -339,9 +337,7 @@ func (svr *Service) loopLoginUntilSuccess(maxInterval time.Duration, firstLoginE
 			Factor:      2,
 			Jitter:      0.1,
 			MaxDuration: maxInterval,
-		}),
-		true,
-		wait.MergeAndCloseOnAnyStopChannel(svr.ctx.Done(), successCh))
+		}), true, svr.ctx.Done())
 }
 
 func (svr *Service) UpdateAllConfigurer(proxyCfgs []v1.ProxyConfigurer, visitorCfgs []v1.VisitorConfigurer) error {

--- a/pkg/util/wait/backoff.go
+++ b/pkg/util/wait/backoff.go
@@ -113,7 +113,7 @@ func (f *fastBackoffImpl) Backoff(previousDuration time.Duration, previousCondit
 	return f.options.Duration
 }
 
-func BackoffUntil(f func() error, backoff BackoffManager, sliding bool, stopCh <-chan struct{}) {
+func BackoffUntil(f func() (bool, error), backoff BackoffManager, sliding bool, stopCh <-chan struct{}) {
 	var delay time.Duration
 	previousError := false
 
@@ -131,7 +131,9 @@ func BackoffUntil(f func() error, backoff BackoffManager, sliding bool, stopCh <
 			delay = backoff.Backoff(delay, previousError)
 		}
 
-		if err := f(); err != nil {
+		if done, err := f(); done {
+			return
+		} else if err != nil {
 			previousError = true
 		} else {
 			previousError = false
@@ -170,9 +172,9 @@ func Jitter(duration time.Duration, maxFactor float64) time.Duration {
 }
 
 func Until(f func(), period time.Duration, stopCh <-chan struct{}) {
-	ff := func() error {
+	ff := func() (bool, error) {
 		f()
-		return nil
+		return false, nil
 	}
 	BackoffUntil(ff, BackoffFunc(func(time.Duration, bool) time.Duration {
 		return period

